components/raftstore-v2/src/operation/ready/apply_trace.rs

```diff
@@ -544,6 +555,74 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
         }
         apply_trace.maybe_advance_admin_flushed(apply_index);
     }
+
+    pub fn flush_before_close<T>(&mut self, ctx: &StoreContext<EK, ER, T>, tx: SyncSender<()>) {
+        info!(
+            self.logger,
+            "region flush before close begin";
+        );
+        let region_id = self.region_id();
+        let flush_threshold: u64 = (|| {
+            fail_point!("flush_before_cluse_threshold", |t| {
+                t.unwrap().parse::<u64>().unwrap()
+            });
+            50
+        })();
+
+        if let Some(tablet) = self.tablet().cloned() {
+            // flush the oldest cf one by one until we are under the replay count threshold
+            loop {
+                let replay_count = self.storage().estimate_replay_count();
+                if replay_count < flush_threshold {
+                    break;
+                }
+                info!(
+                    self.logger,
+                    "flush-before-close: replay count exceeds threshold, pick the oldest cf to flush";
+                    "count" => replay_count,
+                );
+                tablet.flush_oldest_cf(true, None).unwrap();
+
+                let flush_state = self.flush_state().clone();
+                let mut apply_trace = self.storage_mut().apply_trace_mut();
+                let mut max_flush_index = 0;
+
+                let flushed_indexes = flush_state.as_ref().flushed_index();
+                for i in 0..flushed_indexes.len() {
+                    let flush_index = flushed_indexes[i].load(Ordering::SeqCst);
+                    let cf = offset_to_cf(i);
+                    apply_trace.on_flush(cf, flush_index);
+                    max_flush_index = u64::max(max_flush_index, flush_index);
+                }
+
+                apply_trace.maybe_advance_admin_flushed(max_flush_index);
+                let admin_flush = apply_trace.admin.flushed;
+                apply_trace.persisted_applied = admin_flush;
+
+                if self.storage().estimate_replay_count() < flush_threshold {
```

Comment:
it can break direct if this condition is true.

Reply:
Yes